Week2/HW02.hs

```diff
@@ -0,0 +1,101 @@
+
+-- A peg can be one of six colors
+data Peg = Red | Green | Blue | Yellow | Orange | Purple
+         deriving (Show, Eq, Ord)
+
+-- A code is defined to simply be a list of Pegs
+type Code = [Peg]
+
+-- A move is constructed using a Code and two integers; the number of
+-- exact matches and the number of regular matches
+data Move = Move Code Int Int
+          deriving (Show, Eq)
+
+-- List containing all of the different Pegs
+colors :: [Peg]
+colors = [Red, Green, Blue, Yellow, Orange, Purple]
+
+-- Exercise 1 -----------------------------------------
+
+-- Get the number of exact matches between the actual code and the guess
+
+-- Example: exactMatches [Red, Blue, Green, Yellow] [Blue, Green, Yellow, Red] == 0
+-- Example: exactMatches [Red, Blue, Green, Yellow] [Red, Purple, Green, Orange] == 2
+exactMatches :: Code -> Code -> Int
+exactMatches [] [] = 0 
+exactMatches [] _ = 0
+exactMatches _ [] = 0
+exactMatches (c1:xs) (c2:ys) = (if c1 == c2 then 1 else 0) + exactMatches xs ys
+
+-- Exercise 2 -----------------------------------------
+
+-- For each peg in xs, count how many times is occurs in ys
+countColors :: Code -> [Int]
+countColors xs = map (\c -> length (filter (== c) xs)) colors
+
+-- helper function for matches
+getMatchCount :: [Int] -> [Int] -> Int
+getMatchCount [] [] = 0
+getMatchCount _ [] = 0
+getMatchCount [] _ = 0
+getMatchCount (x:xs) (y:ys) = (if x>0 && y>0 then min x y else 0) + getMatchCount xs ys
+
+-- Count number of matches between the actual code and the guess
+matches :: Code -> Code -> Int
+matches [] [] = 0
+matches [] _ = 0
+matches _ [] = 0
+matches (c1:c1x) (c2:c2x) = getMatchCount (countColors (c1:c1x)) (countColors (c2:c2x))
+
+-- Exercise 3 -----------------------------------------
+
+-- Construct a Move from a guess given the secret
+getMove :: Code -> Code -> Move
+getMove sec move = Move move exMatch (matches sec move - exMatch)
+    where   exMatch = exactMatches sec move
+
+-- Exercise 4 -----------------------------------------
+
+isConsistent :: Move -> Code -> Bool
+isConsistent mv@(Move guess _ _) code = mv == getMove code guess
+
+-- Exercise 5 -----------------------------------------
+
+filterCodes :: Move -> [Code] -> [Code]
+filterCodes = filter . isConsistent 
+
+-- Exercise 6 -----------------------------------------
+
+-- I wasnt able to solve this one on my own 
+
+allCodesHelper :: Code -> [Code]
+allCodesHelper code = map (:code) colors
+
+allCodes :: (Code -> [Code]) -> Int -> [Code]
+allCodes _          0 = [[]]
+allCodes addColorFn n
+  | n > 0     = concatMap addColorFn nxtCodes
+  | otherwise = [[]]
+  where
+    nxtCodes  = allCodes addColorFn $ n-1  
+                  
+
+-- Exercise 7 -----------------------------------------
+solveHelper :: Code -> [Code] -> [Move]
+solveHelper _                    [] = []
+solveHelper secret (guess:remCodes) = move: nextAllowedCodes
+  where
+    move               = getMove secret guess
```

Comment:
You don't compare at all if the move is any good.
E.g the move you pick might have 0 matches, but you just accept it and filter remaining codes to be compatible with this 0-match move. You won't find any better moves in the list of moves compatible with 0-match move.

Reply:
I think what you meant was that the algorithm is "dumb" which it is supposed to be, hahaha, nevertheless it'll still converge to one correct answer as per the exercise. 

> This txt is from the pdf in Exercise 7

![Screenshot 2021-01-11 at 10 30 09 PM](https://user-images.githubusercontent.com/30658517/104213668-9183f200-545c-11eb-874d-191107c821e4.png)

Let me know if I understood ur review wrong or if its the same thing.
Thanks again for pointing it out specifically

Reply:
Oh, right, I didn't realize that. I was missing some kind of more aggressive pruning of the possibilities, but now I see it will still converge.